modules/fedimint-wallet-client/src/deposit.rs

```diff
@@ -0,0 +1,373 @@
+    }
+}
+
+async fn await_created_btc_transaction_submitted(
+    context: WalletClientContext,
+    tweak: KeyPair,
+) -> (bitcoin::Transaction, u32) {
+    let script = context
+        .wallet_descriptor
+        .tweak(&tweak.public_key().to_x_only_pubkey(), &context.secp)
+        .script_pubkey();
+    loop {
+        match get_script_history(&context.esplora_server, &script).await {
+            Ok(received) => {
+                // TODO: fix
+                if received.len() > 1 {
+                    warn!("More than one transaction was sent to deposit address, only considering the first one");
+                }
+
+                if let Some(transaction) = received.into_iter().next() {
+                    let out_idx = transaction
+                        .output
+                        .iter()
+                        .enumerate()
+                        .find_map(|(idx, output)| {
+                            if output.script_pubkey == script {
+                                Some(idx as u32)
+                            } else {
+                                None
+                            }
+                        })
+                        .expect("TODO: handle invalid tx returned by API");
+
+                    return (transaction, out_idx);
+                } else {
+                    trace!("No transactions received yet for script {script:?}");
+                }
+            }
+            Err(e) => {
+                warn!("Error fetching transaction history for {script:?}: {e}");
+            }
+        }
+
+        sleep(TRANSACTION_STATUS_FETCH_INTERVAL).await;
+    }
+}
+
+async fn get_script_history(
+    esplora_server: &str,
+    script: &Script,
+) -> anyhow::Result<Vec<bitcoin::Transaction>> {
+    let esplora_client = esplora_client(esplora_server)?;
+    let script = script.clone();
+
+    let transactions = esplora_client
+        .scripthash_txs(&script, None)
+        .await?
+        .into_iter()
+        .map(|tx| tx.to_tx())
+        .collect::<Vec<_>>();
+
+    Ok(transactions)
+}
+
+async fn transition_tx_seen(
+    old_state: DepositStateMachine,
+    btc_transaction: bitcoin::Transaction,
+    out_idx: u32,
+) -> DepositStateMachine {
+    let DepositStateMachine {
+        operation_id,
+        state: old_state,
+    } = old_state;
+
+    match old_state {
+        DepositStates::Created(created_state) => DepositStateMachine {
+            operation_id,
+            state: DepositStates::WaitingForConfirmations(WaitingForConfirmationsDepositState {
+                tweak_key: created_state.tweak_key,
+                btc_transaction,
+                out_idx,
+            }),
+        },
+        state => panic!("Invalid previous state: {state:?}"),
+    }
+}
+
+async fn await_deposit_address_timeout(timeout_at: SystemTime) {
+    if let Ok(time_until_deadline) = timeout_at.duration_since(fedimint_core::time::now()) {
+        tokio::time::sleep(time_until_deadline).await;
```

Comment:
This will break wasm. Use `fedimint_core::task::sleep` instead.

Reply:
Done. Wonder if we could just use semgrep to ban it …

Reply:
https://github.com/fedimint/fedimint/pull/2564